src/circuit/mont.rs

```diff
@@ -0,0 +1,445 @@
+
+        // y' + y = lambda(x - x')
+        cs.enforce(
+            || "evaluate yprime",
+            LinearCombination::zero()           + self.x.get_variable()
+                                                - xprime.get_variable(),
+
+            LinearCombination::zero()           + lambda.get_variable(),
+
+            LinearCombination::<Var, E>::zero() + yprime.get_variable()
+                                                + self.y.get_variable()
+        );
+
+        Ok(MontgomeryPoint {
+            x: xprime,
+            y: yprime
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use bellman::{ConstraintSystem};
+    use rand::{XorShiftRng, SeedableRng, Rng};
+    use pairing::bls12_381::{Bls12, Fr};
+    use pairing::{Field};
+    use ::circuit::test::*;
+    use ::jubjub::{
+        montgomery,
+        JubjubBls12
+    };
+    use super::{MontgomeryPoint, AllocatedNum, Boolean};
+    use super::super::boolean::AllocatedBit;
+    use ::group_hash::group_hash;
+
+    #[test]
+    fn test_group_hash() {
+        let params = &JubjubBls12::new();
+        let rng = &mut XorShiftRng::from_seed([0x3dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
+
+        let mut num_errs = 0;
+        let mut num_unsatisfied = 0;
+        let mut num_satisfied = 0;
+
+        for _ in 0..100 {
+            let mut cs = TestConstraintSystem::<Bls12>::new();
+
+            let mut tag_bytes = vec![];
+            let mut tag = vec![];
+            for i in 0..10 {
+                let mut byte = 0;
+                for j in 0..8 {
+                    byte <<= 1;
+                    let b: bool = rng.gen();
+                    if b {
+                        byte |= 1;
+                    }
+                    tag.push(Boolean::from(
+                        AllocatedBit::alloc(
+                            cs.namespace(|| format!("bit {} {}", i, j)),
+                            Some(b)
+                        ).unwrap()
+                    ));
+                }
+                tag_bytes.push(byte);
+            }
+
+            let p = MontgomeryPoint::group_hash(
+                cs.namespace(|| "gh"),
+                &tag,
+                params
+            );
+
+            let expected = group_hash::<Bls12>(&tag_bytes, params);
+
+            if p.is_err() {
+                assert!(expected.is_none());
+                num_errs += 1;
+            } else {
+                if !cs.is_satisfied() {
+                    assert!(expected.is_none());
+                    num_unsatisfied += 1;
+                } else {
+                    let p = p.unwrap();
+                    let (x, y) = expected.unwrap();
+
+                    assert_eq!(p.x.get_value().unwrap(), x);
+                    assert_eq!(p.y.get_value().unwrap(), y);
+
+                    num_satisfied += 1;
+                }
+            }
+        }
+
+        assert_eq!(
+            (num_errs, num_unsatisfied, num_satisfied),
+            (47, 4, 49)
```

Comment:
I'm not particularly keen on tests that just count the number of errors/satisfied/unsatisfied in a pseudorandom sample, because:

* these numbers are inevitably just found by running the implementation, so they're not an independent check, only a regression test
* even as a regression test, it's not entirely reliable because in principle two changes could cancel out in the totals
* it's difficult to debug failures.

Reply:
I agree. This is just a temporary alternative to more thorough test vectors.